### PR TITLE
Feature/1.16 protocol

### PR DESCRIPTION
**Pull request by bundabrg**
WIP on serializers etc.

## Todo
* [x] AvailableCommandsPacket
* [x] CraftingDataPacket
* [x] PlayerListPacket
* [x] StartGamePacket
* [x] InventorySlotPacket
* [x] SetSpawnPositionPacket
* [x] InventoryContentPacket
* [x] CreativeContentPacket
* [x] InventoryTransactionPacket
* [x] AdventureSettings 
* [x] EntityEvent 
* [ ] Get ProxyPass working with serialized packets instead of wrapped
* [x] EntityLinkPacket
* [x] EventPacket
* [x] ItemStackRequestPacket
* [x] ItemStackResponsePacket

## Binary Incompatible Packets
I compare the binary of incoming and outgoing packets and if they differ they go in this list. Some are because of ordering of
elements but it may be worth ensuring ordering is preserved now. A dump of each packet (orginal/encoded) will be in the NextCloud folder. These don't necessarily indicate breaking issues. Also I'm more interested in packets of different size since
out of order is not something worth fixing if it aint broke.

* [x] AddPlayerPacket
* ~~AvailableCommandsPacket~~ - This packet loses unused enums, so probably not worth it.
* ~~AvailableEntityIdentifiersPacket~~ - Just out of order NBT tag which doesn't need fixing.
* ~~BiomeDefinitionsListPacket~~ - Out of order NBT tag. wontfix
* ~~CraftingDataPacket~~ - Fine apart from NBT tag ordering
* [x] ~~CreativeContentPacket~~ - Retain sorting for this. Only NBT out of order now.
* [x] LevelEventPacket
* ~~InventoryContentPacket~~ - May need a rethink or maybe working better than vanilla.
* ~~SetEntityDataPacket~~ Flags and Flags2 are put even if either are empty. Not worth fixing.
* ~~SetSpawnPositionPacket~~ - Will be fixed once https://github.com/NukkitX/Network/pull/10 is merged into master
* ~~StartGamePacket~~ - Needs https://github.com/NukkitX/Network/pull/10 but also has NBT tags.

## Tidy up
* [x] StartGamePacket
* [x] Remove debug stuff

## Notes
* StartGamePacket is now deserializing but there are a few unknowns.
* https://cloud.grieve.com.au/s/FJYzFHeBoQPMxFj - Contains dumps of the packets, plus my work in a spreadsheet mapping stuff out plus the dumped proxypass data.

### CODE REVIEW

**Comment by bundabrg** [comment]
Add to my list

File: bedrock/bedrock-common/src/main/java/com/nukkitx/protocol/bedrock/packet/SetSpawnPositionPacket.java [hunk 1]
```diff
@@ -10,8 +10,25 @@
 @Data
 @EqualsAndHashCode(doNotUseGetters = true, callSuper = false)
 public class SetSpawnPositionPacket extends BedrockPacket {
+    // spawnType is the type of spawn to set. It is either PLAYER_SPAWN or WORLD_SPAWN, and specifies
+    // the behaviour of the spawn set. If WORLD_SPAWN is set, the position to which compasses will point is
+    // also changed.
```
Comment (SupremeMortal) [inline bedrock/bedrock-common/src/main/java/com/nukkitx/protocol/bedrock/packet/SetSpawnPositionPacket.java:15]: Could you use proper Javadoc annotations here?

File: bedrock/bedrock-common/src/main/java/com/nukkitx/protocol/bedrock/packet/CreativeContentPacket.java [hunk 2]
```diff
@@ -2,20 +2,26 @@
...
+ * creative inventory window ID.
+ */
 @Data
 @EqualsAndHashCode(doNotUseGetters = true, callSuper = false)
 public class CreativeContentPacket extends BedrockPacket {
     /**
      * Item entries for the creative menu. Each item must have a unique ID for the net ID manager
      */
-    private final Int2ObjectMap<ItemData> entries = new Int2ObjectOpenHashMap<>();
+    private final Int2ObjectMap<ItemDataInstance> entries = new Int2ObjectOpenHashMap<>();
```
Comment (SupremeMortal) [inline bedrock/bedrock-common/src/main/java/com/nukkitx/protocol/bedrock/packet/CreativeContentPacket.java:24]: I'm confused why this was changed. The network ID is the map key so the extra class seems a bit pointless.

Comment (bundabrg) [reply bedrock/bedrock-common/src/main/java/com/nukkitx/protocol/bedrock/packet/CreativeContentPacket.java:24]: I wasn't sure on that one, but looking at other packets (like SetSlotPacket) they all have networkID's as well so it seemed an idea to move it to an instance. However if you prefer to flatten that structure into the packets then that can be changed back.

File: bedrock/bedrock-common/src/main/java/com/nukkitx/protocol/bedrock/packet/SetSpawnPositionPacket.java (hunk 1 above)
Comment (bundabrg) [reply bedrock/bedrock-common/src/main/java/com/nukkitx/protocol/bedrock/packet/SetSpawnPositionPacket.java:15]: Can you give me an example? As far as I know there is no such thing as javadoc for class variables unless you run this through delombok and run javadoc over that?

File: bedrock/bedrock-common/src/main/java/com/nukkitx/protocol/bedrock/packet/CreativeContentPacket.java (hunk 2 above)
Comment (lukeeey) [reply bedrock/bedrock-common/src/main/java/com/nukkitx/protocol/bedrock/packet/CreativeContentPacket.java:24]: Is the network id not included in the ItemData class though?

Comment (bundabrg) [reply bedrock/bedrock-common/src/main/java/com/nukkitx/protocol/bedrock/packet/CreativeContentPacket.java]: I don't think so as its a runtime ID held by the server for that specific instance of itemstack and some itemstacks may not have it. I think needs to be separate either as an index on the collection like SupremeMortal has or a separate object. 
I've flattened the structure and worked how to provide backwards compatibility in the InventoryContentsPacket (my main reason for the change) so this should be resolved now.

File: bedrock/bedrock-v407/src/main/java/com/nukkitx/protocol/bedrock/v407/serializer/StartGameSerializer_v407.java [hunk 3]
```diff
@@ -36,6 +36,12 @@ public void serialize(ByteBuf buffer, BedrockPacketHelper helper, StartGamePacke
         helper.writeVector2f(buffer, packet.getRotation());
         // Level settings start
         VarInts.writeInt(buffer, packet.getSeed());
+
+        //@Todo
+        buffer.writeByte(0);
```
Comment (SupremeMortal) [inline bedrock/bedrock-v407/src/main/java/com/nukkitx/protocol/bedrock/v407/serializer/StartGameSerializer_v407.java:41]: SpawnSetttings:
ShortLE - SpawnBiomeType: 0 = `DEFAULT`, 1 =`USER_DEFINED`
String - User defined biome name
VarInt - dimension

File: bedrock/bedrock-v407/src/main/java/com/nukkitx/protocol/bedrock/v407/serializer/StartGameSerializer_v407.java [hunk 4]
```diff
@@ -45,9 +51,13 @@ public void serialize(ByteBuf buffer, BedrockPacketHelper helper, StartGamePacke
         VarInts.writeInt(buffer, packet.getDayCycleStopTime());
         VarInts.writeInt(buffer, packet.getEduEditionOffers());
         buffer.writeBoolean(packet.isEduFeaturesEnabled());
-        helper.writeString(buffer, packet.getUnknownString0());
+//        helper.writeString(buffer, packet.getUnknownString0());
```
Comment (SupremeMortal) [inline bedrock/bedrock-v407/src/main/java/com/nukkitx/protocol/bedrock/v407/serializer/StartGameSerializer_v407.java:54]: This should be here. `Education product ID`

File: bedrock/bedrock-v407/src/main/java/com/nukkitx/protocol/bedrock/v407/serializer/StartGameSerializer_v407.java [hunk 5]
```diff
@@ -71,6 +81,10 @@ public void serialize(ByteBuf buffer, BedrockPacketHelper helper, StartGamePacke
         buffer.writeIntLE(packet.getUnknownInt0());
         buffer.writeIntLE(packet.getUnknownInt1());
```
Comment (SupremeMortal) [inline bedrock/bedrock-v407/src/main/java/com/nukkitx/protocol/bedrock/v407/serializer/StartGameSerializer_v407.java:82]: Limited world width
Limited world depth

File: bedrock/bedrock-v407/src/main/java/com/nukkitx/protocol/bedrock/v407/serializer/StartGameSerializer_v407.java [hunk 6]
```diff
@@ -71,6 +81,10 @@ public void serialize(ByteBuf buffer, BedrockPacketHelper helper, StartGamePacke
...
 
+        // @Todo
+        buffer.writeByte(0);
+        buffer.writeByte(0);
```
Comment (SupremeMortal) [inline bedrock/bedrock-v407/src/main/java/com/nukkitx/protocol/bedrock/v407/serializer/StartGameSerializer_v407.java:86]: bool - nether type
bool - force experimental gameplay

**Comment by Sleepybear** [comment]
> Added to my list

I went ahead and updated it and did a PR :)

**Comment by Sleepybear** [comment]
Updated Inventory Transaction and ContainerType enum and made a PR https://github.com/bundabrg/Protocol/pull/1

File: bedrock/bedrock-v407/src/main/java/com/nukkitx/protocol/bedrock/v407/serializer/StartGameSerializer_v407.java (hunk 5 above)
Comment (bundabrg) [reply bedrock/bedrock-v407/src/main/java/com/nukkitx/protocol/bedrock/v407/serializer/StartGameSerializer_v407.java:82]: depth or height?

File: bedrock/bedrock-v407/src/main/java/com/nukkitx/protocol/bedrock/v407/serializer/StartGameSerializer_v407.java (hunk 6 above)
Comment (bundabrg) [reply bedrock/bedrock-v407/src/main/java/com/nukkitx/protocol/bedrock/v407/serializer/StartGameSerializer_v407.java:86]: nether type is a bool?

**Comment by D3ATHBRINGER13** [comment]
How do you even find out which packets to update?
It must be sorcery

**Comment by bundabrg** [comment]
Trial and error. If it doesn't work it'll throw an exception. Its the ones that quietly seem to work but don't that are the tricky ones.

**Comment by bundabrg** [comment]
As of this point running Bedrock 1.16 through Geyser to a 1.16 server no longer throws errors anymore.

**Comment by bundabrg** [comment]
Actually there is still some issues. If I force ProxyPass to send serialized packets instead of the wrapped ones then pressing pause crashes my client. So likely the creative inventory packet is not actually sending properly.

File: bedrock/bedrock-v407/src/main/java/com/nukkitx/protocol/bedrock/v407/serializer/StartGameSerializer_v407.java (hunk 6 above)
Comment (SupremeMortal) [reply bedrock/bedrock-v407/src/main/java/com/nukkitx/protocol/bedrock/v407/serializer/StartGameSerializer_v407.java]: Probably means old vs new nether

**Comment by bundabrg** [comment]
```
 Original  : ClientCacheMissResponsePacket(blobs={-5088169596681128977=>[B@63427b6f, 5525890368274462595=>[B@77704124, 697244647362271770=>[B@43603004, 6277142989889253390=>[B@4f4184f9, -2141166721229966853=>[B@3f28177d, 5344797489550220338=>[B@4aa418dd, 7903962874019228013=>[B@806e19d, -4284324402056478644=>[B@4b23ca28, 1926365300632061930=>[B@344fafce, 3519612410177071169=>[B@5e472b6a, -2464178785407799018=>[B@61723336, 5650810877501159970=>[B@3455d7ba, 4867916228362537832=>[B@7dad2f21, -1551326375716972070=>[B@1a564785, -8782755279894052686=>[B@6acd3c7a, -2460236847038805664=>[B@703fce9e, 5382989540730338675=>[B@64ff239f, -738510946314022665=>[B@4cbd3aea, -5392095405093548993=>[B@4f10e1eb, -5676988546385803833=>[B@41559da9, -6769769296720347836=>[B@5cabadc7, -3687877336267152641=>[B@79e4a6d2, 1453149995863778557=>[B@31330fbc, -6329307095232152075=>[B@5c6674e6, -6020612170357560079=>[B@5da22f38, -8729834000600695867=>[B@12cc068, 4110668932288691043=>[B@284b611e, 5163442875421172632=>[B@47e5e1b5, -2019741700444270559=>[B@b72dd9e, -5584127321672978268=>[B@6c3a7318, 9042939185607409177=>[B@1f0aac10, 657589507032669295=>[B@778a0e0d, 6056293070384433371=>[B@6a89b34f, 8264390922811756709=>[B@5594bfb5, 5742430771134339985=>[B@1fb4f33b, 3920767459076044935=>[B@38696ca6, -839287375603888905=>[B@1fddb32f, 6335569569637433452=>[B@6f44de76, -8410020589298533014=>[B@57d1d58e, -6558828935063435206=>[B@24ae9f8e, 2975627062405142200=>[B@569804ff, 5306489147653416722=>[B@1f66c734, 9195373650001222665=>[B@46420cf3, 8980563227405757831=>[B@16c0a95a, 7698047548056232783=>[B@69e881, -2384956916163549369=>[B@299fd0df, 2716815044098525580=>[B@7f6c748b, 2921718564612889376=>[B@6cf58845, 1316619653994133605=>[B@239748ba, 5026057497673339104=>[B@74d97d3f, 4644425080732242547=>[B@680d9703})
Re-encoded : ClientCacheMissResponsePacket(blobs={-5088169596681128977=>[B@287a8de7, 5525890368274462595=>[B@3131d2ca, 697244647362271770=>[B@ebe6156, 6277142989889253390=>[B@a0f4a0d, -2141166721229966853=>[B@7666fb1e, 5344797489550220338=>[B@497ec09a, 7903962874019228013=>[B@34b7ac32, -4284324402056478644=>[B@4418d78b, 1926365300632061930=>[B@5677a1d, 3519612410177071169=>[B@3aa2103b, 5650810877501159970=>[B@33fded61, -2464178785407799018=>[B@315bc01a, 4867916228362537832=>[B@fd2e3be, -8782755279894052686=>[B@255b09b7, -1551326375716972070=>[B@79116e98, -2460236847038805664=>[B@16a52ab0, 5382989540730338675=>[B@ef2e4c9, -738510946314022665=>[B@73994211, -5392095405093548993=>[B@10e8754, -5676988546385803833=>[B@3e41cd6f, -6769769296720347836=>[B@270cb2c2, -3687877336267152641=>[B@1e3ce5f8, 1453149995863778557=>[B@6bfdf54c, -6020612170357560079=>[B@5be777e9, -6329307095232152075=>[B@19bf54f5, -8729834000600695867=>[B@6055611f, 4110668932288691043=>[B@6406bfcc, 5163442875421172632=>[B@4bec49a, -2019741700444270559=>[B@67452004, -5584127321672978268=>[B@2c53f86a, 9042939185607409177=>[B@7d7acaf3, 657589507032669295=>[B@f124bd2, 6056293070384433371=>[B@41991f34, 8264390922811756709=>[B@723a2fc5, 5742430771134339985=>[B@59bc5e23, 3920767459076044935=>[B@44edd21d, -839287375603888905=>[B@4d04150b, 6335569569637433452=>[B@6ee35f69, -8410020589298533014=>[B@559549d0, -6558828935063435206=>[B@7f421445, 2975627062405142200=>[B@1283faeb, 5306489147653416722=>[B@666ff7c5, 9195373650001222665=>[B@19cbf911, 8980563227405757831=>[B@76443651, 7698047548056232783=>[B@7b1dded0, -2384956916163549369=>[B@6255cfa9, 2716815044098525580=>[B@2490b183, 2921718564612889376=>[B@73b1edfe, 1316619653994133605=>[B@664f5606, 5026057497673339104=>[B@2aabf5e1, 4644425080732242547=>[B@b96559e})
```

**Comment by bundabrg** [comment]
As ordering in some packets are now becoming important I've added a new section to my todo of something I wanted to do for a little while and that is ensure the encoded bytes from a packet are the same as a decoded packet. This may catch some other errors but will also catch anything that is out of order (and thus probably not a breaking issue but still nice to get right).

**Comment by bundabrg** [comment]
If we receive the following inventory content packet:

```
7c 35 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
```

when we re-serialize it it comes to:
```
7c 01 00 00
```

The reason being the networkID is 0 for every item in the first and thus we lose out on the data as its being stored with the networkId as a reference. Now this is probably ok since each item should have a unique networkId but its something to be aware of.

**Pull request merged**